### PR TITLE
Rework of PillarDAO

### PR DESCRIPTION
**Pull request by cryptonoyaiba**
- Allows for passing in at deployment an existing array of accounts (for our specific use case). This will allow for any members that already exist to be migrated onto a new DAO contract. This expects a MembershipNFT contract to already be in use with an existing set of NFTs minted to those members.
- Addition of deployment script with setup for minting new members to existing MembershipNFT contract.
- Added test to mimic entire process through to withdrawal.
- Upgraded hardhat version to 2.13.0 and new hardhat-network-helpers package for testing time dependant tests.

### CODE REVIEW

File: contracts/PillarDAO.sol [hunk 1]
```diff
@@ -127,32 +128,52 @@ contract PillarDAO is IPillarDAO, Ownable, ReentrancyGuard {
...
-    //     });
-    //     memberships[0xb1E6C220925bb475C694E896645f5636C0D019dc] = 5;
-    //     balances[0xb1E6C220925bb475C694E896645f5636C0D019dc] = Deposit({
-    //         depositAmount: 10000 ether,
-    //         depositTime: block.timestamp
-    //     });
-    // }
+    function _addExistingMembers(address[9] memory _members) internal {
+        // pre-add existing DAO members [9]
+        memberships[_members[0]] = 1;
```
Comment (ch4r10t33r) [inline contracts/PillarDAO.sol]: Should we add a validation to ensure the address is not 0?

File: contracts/PillarDAO.sol [hunk 2]
```diff
@@ -127,32 +128,52 @@ contract PillarDAO is IPillarDAO, Ownable, ReentrancyGuard {
...
+        });
+        memberships[_members[1]] = 2;
+        balances[_members[1]] = Deposit({
+            depositAmount: 10000 ether,
+            depositTime: block.timestamp
+        });
+        memberships[_members[2]] = 3;
+        balances[_members[2]] = Deposit({
+            depositAmount: 10000 ether,
+            depositTime: block.timestamp
```
Comment (ch4r10t33r) [inline contracts/PillarDAO.sol]: Shouldn't we retain the original timestamp of the membership?

File: package.json [hunk 3]
```diff
@@ -1,13 +1,14 @@
 {
   "name": "pillar-dao-contracts",
   "devDependencies": {
+    "@nomicfoundation/hardhat-network-helpers": "^1.0.9",
```
Comment (ch4r10t33r) [inline package.json]: Pls stick to a specific version of the package.

**Review by ch4r10t33r — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (cryptonoyaiba) [reply package.json]: done

File: contracts/PillarDAO.sol (hunk 1 above)
Comment (cryptonoyaiba) [reply contracts/PillarDAO.sol]: Added validation for `address(0)`. Also changed this from `address[9]` to a dynamic length incase we need to reuse and make `_addExistingMembers` a loop to account for dynamic length. Added test in suite to confirm revert occurs.

File: contracts/PillarDAO.sol (hunk 2 above)
Comment (cryptonoyaiba) [reply contracts/PillarDAO.sol]: @ch4r10t33r The only way I could do this is look when the `deposit` tx went through and convert that to a UNIX timestamp. There is no way to view that info from the contract. That would require me to change the shape of the data passed in. I can do this if you are happy with that approach?

Comment (ch4r10t33r) [reply contracts/PillarDAO.sol]: Cant we pass the timestamps also as an additional array?

**Comment by cryptonoyaiba** [comment]
Set up test environment for QA (migration completed) on Mumbai.

**Review by ch4r10t33r — APPROVED** [review]

**Pull request merged**